utils/value_or_reference.hh

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+template <typename T>
+struct value_or_reference {
+    std::optional<T> _opt;
+    const T& _ref;
+
+    value_or_reference(T&& v) : _opt(std::move(v)), _ref(*_opt) {}
```

Comment:
Agreed.
Also, if we ever need it to be default constructed we may want to keep a pointer rather than a refernce member and initialize it to nullptr.

Reply:
It was extracted from `sstables/mx/reader.cc` exactly as it was written. The goal was to unify the interfaces between readers. Got into a segfault because I did not expect a difference. Adding `explicit` will also increase readability so definitely to be added.